dotCMS/src/main/java/com/dotcms/ema/EMAWebInterceptor.java

```diff
@@ -0,0 +1,208 @@
+import com.dotmarketing.business.web.WebAPILocator;
+import com.dotmarketing.exception.DotRuntimeException;
+import com.dotmarketing.util.Logger;
+import com.dotmarketing.util.PageMode;
+import com.dotmarketing.util.UtilMethods;
+import com.dotmarketing.util.WebKeys;
+import com.dotmarketing.util.json.JSONObject;
+import com.google.common.collect.ImmutableMap;
+import com.liferay.portal.model.User;
+
+public class EMAWebInterceptor  implements WebInterceptor{
+
+    public  static final String      PROXY_EDIT_MODE_URL_VAR = "proxyEditModeURL";
+    private static final String      API_CALL                = "/api/v1/page/render";
+    public static final String EMA_APP_CONFIG_KEY = "dotema-config";
+    private static final ProxyTool   proxy                   = new ProxyTool();
+
+
+    @Override
+    public String[] getFilters() {
+        return new String[] {
+                API_CALL + "*"
+        };
+    }
+
+    @Override
+    public Result intercept(final HttpServletRequest request, final HttpServletResponse response) throws IOException {
+
+        final Host currentHost = getCurrentHost(request);
+
+        if (!this.existsConfiguration(currentHost.getIdentifier())) {
+            return Result.NEXT;
+        }
+
+        final Optional<String> proxyUrl = proxyUrl(currentHost);
+        final PageMode mode             = PageMode.get(request);
+
+        if (!proxyUrl.isPresent() || mode == PageMode.LIVE) {
+            return Result.NEXT;
+        }
+
+        Logger.info(this.getClass(), "GOT AN EMA Call --> " + request.getRequestURI());
+
+        return new Result.Builder().wrap(new MockHttpCaptureResponse(response)).next().build();
+    }
+
+
+    @Override
+    public boolean afterIntercept(final HttpServletRequest request, final HttpServletResponse response) {
+
+        try {
+
+            if (response instanceof MockHttpCaptureResponse) {
+
+                final Host currentHost = getCurrentHost(request);
+                final Optional<String> proxyUrl            = proxyUrl(currentHost);
+                final MockHttpCaptureResponse mockResponse = (MockHttpCaptureResponse)response;
+                final String postJson                      = new String(mockResponse.getBytes());
+                final JSONObject json                      = new JSONObject(postJson);
+                final Map<String, String> params           = ImmutableMap.of("dotPageData", postJson);
+                
+                Logger.info(this.getClass(), "Proxying Request --> " + proxyUrl.get());
+                
+                String responseStr = new String();
+                final ProxyResponse pResponse = proxy.sendPost(proxyUrl.get(), params);
+
+                if (pResponse.getResponseCode() == 200) {
+                    responseStr = new String(pResponse.getResponse());
+                }else {
+                    responseStr+="<html><body>";
+                    responseStr+="<h3>Unable to connect with the rendering engine</h3>";
```

Comment:
Codacy found an issue: [Prefer StringBuilder (non-synchronized) or StringBuffer (synchronized) over += for concatenating strings](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5922984)